numpy/_core/src/multiarray/descriptor.c

```diff
@@ -1813,12 +1813,15 @@ _convert_from_str(PyObject *obj, int align)
         int kind;
 
         /* Parse the integer, make sure it's the rest of the string */
-        elsize = (int)strtol(type + 1, &typeend, 10);
-        /* Make sure size is not negative */
-        if (elsize < 0) {
+        long result = strtol(type + 1, &typeend, 10);
+        // make sure it doesn't overflow or go negative
+        if (result > INT_MAX || result < 0) {
             goto fail;
         }
 
+        elsize = (int)result;
+
+
         if (typeend - type == len) {
```

Comment:
Shouldn't an error be raised if this is not true?

Reply:
Right now it falls through to `goto fail` eventually but I can see if anything breaks if add an `else {goto fail;}` block.